tests/weak_forms/step-8-variant_01c_vec.cc

```diff
@@ -0,0 +1,157 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 - 2022 by Jean-Paul Pelteret
+//
+// This file is part of the Weak forms for deal.II library.
+//
+// The Weak forms for deal.II library is free software; you can use it,
+// redistribute it, and/or modify it under the terms of the GNU Lesser
+// General Public License as published by the Free Software Foundation;
+// either version 3.0 of the License, or (at your option) any later
+// version. The full text of the license can be found in the file LICENSE
+// at the top level of the Weak forms for deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// Elasticity problem: Assembly using weak forms
+// This variant directly uses the scalar coefficients in the weak form.
+// This test replicates step-8 exactly.
+// - Vectorized variant
+
+#include <deal.II/base/function.h>
+
+#include <weak_forms/weak_forms.h>
+
+#include "../weak_forms_tests.h"
+#include "wf_common_tests/step-8.h"
+
+
+using namespace dealii;
+
+
+
+template <int dim>
+class Step8 : public Step8_Base<dim>
+{
+public:
+  Step8();
+
+protected:
+  void
+  assemble_system() override;
+};
+
+
+template <int dim>
+Step8<dim>::Step8()
+  : Step8_Base<dim>()
+{}
+
+
+template <int dim>
+void
+Step8<dim>::assemble_system()
+{
+  using namespace WeakForms;
+
+  // Symbolic types for test function, trial solution and a coefficient.
+  const TestFunction<dim>          test;
+  const TrialSolution<dim>         trial;
+  const SubSpaceExtractors::Vector subspace_extractor(0, "u", "\\mathbf{u}");
+
+  // Material coefficients
+  const double lambda = 1.0;
+  const double mu     = 1.0;
+
+  const VectorFunctionFunctor<dim> rhs_coeff("s", "\\mathbf{s}");
+  const RightHandSide<dim>         rhs;
+
+  const auto test_ss  = test[subspace_extractor];
+  const auto trial_ss = trial[subspace_extractor];
+
+  const auto test_val   = test_ss.value();
+  const auto test_grad  = test_ss.gradient();
+  const auto trial_grad = trial_ss.gradient();
+
+
+  MatrixBasedAssembler<dim> assembler;
+  assembler +=
+    bilinear_form(test_grad, lambda, trial_grad)
+      .template component_filter<multiplicity_I | dof_I_component_i |
+                                 multiplicity_J | dof_J_component_j>()
+      .dV() +
+    bilinear_form(test_grad, mu, trial_grad)
+      .template component_filter<multiplicity_I | dof_I_component_j |
+                                 multiplicity_J | dof_J_component_i>()
+      .dV() +
+    bilinear_form(test_grad, mu, trial_grad).delta_IJ().dV() -
+    linear_form(test_val, rhs_coeff.value(rhs)).dV();
```

Comment:
Hi @Rombur, I've finally finished a shape component filter. This allows one to write step-8 using scalar material parameters, rather than the full rank-4 material tensor. I'd be happy to get some input on the naming of these enumerations.

Reply:
The `dof_X_component_y` makes perfect sense to me but I am not sure what the `multiplicity_X` represents.

Reply:
I've added some documentation to the enumeration. Please let me know if you find it to be less than adequate, and we can adjust it. I find it hard to explain what it means really (excluding the big table in the intro to the [FiniteElement class](https://dealii.org/current/doxygen/deal.II/classFiniteElement.html). I think that we don't do such a great job ourselves in the documentation), and for me its still difficult to grok what [`fe.system_to_component_index()`](https://www.dealii.org/developer/doxygen/deal.II/classFiniteElement.html#a23200d54e40cd719a183d47b6cfb719f) and [`fe.system_to_base_index()`](https://www.dealii.org/developer/doxygen/deal.II/classFiniteElement.html#ad88459d21a0e6c8b8882eb5dc981cf31) actually return.

Basically, I looked at the requirements of step-8 and then thought about the implications when using an FESystem with different fields, as well as Extractors. That's how I settled on needing the multiplicity to render [this (not using extractors)](https://github.com/jppelteret/dealii-weak_forms/pull/71/files#diff-738acd5907a3594d13b552399cb89f14180b25d1767e5360cdb4a466c0259eb7R151-R153) and [this (with extractors)](https://github.com/jppelteret/dealii-weak_forms/pull/71/files#diff-738acd5907a3594d13b552399cb89f14180b25d1767e5360cdb4a466c0259eb7R199-R204) as equivalent operations. I hope that this is right -- maybe some further tests with coupled systems would fully verify the concept.

Reply:
> I've added some documentation to the enumeration.

I don't see it. Did you forget to push it?

> t. I find it hard to explain what it means really (excluding the big table in the intro to the [FiniteElement class](https://dealii.org/current/doxygen/deal.II/classFiniteElement.html). I think that we don't do such a great job ourselves in the documentation)

Now I understand what you mean. I think referring to deal.II's documentation would be helpful.

Reply:
Bah, you're right. I forgot to push that commit. I've opened https://github.com/jppelteret/dealii-weak_forms/pull/97 with the missing documentation.